content/blog/2025-03-11-Alive-Peephole-Optimizations

```diff
@@ -0,0 +1,62 @@
+
+There were also some concerns about correctness in Alive’s verification procedures, since they rely on the use of SMT solvers, which have known bugs in them. However, again in in-class discussion, we couldn’t come up with any better solution; to paraphrase Prof. Sampson in our in-class discussion, “It seems to me that SMT solvers are the worst solution, besides all the other solutions.”
+
+Alive is designed to work closely alongside LLVM, but they do have a note acknowledging that, while the tool follows the semantics described in the documentation for LLVM, there could be human misunderstandings which create disparities in the actual semantics of LLVM versus the semantics written in the documentation. While it is obviously not Alive’s fault if LLVM, it does introduce another degree of uncertainty about the correctness of Alive.
+
+Overall, while there were some concerns raised about Alive in the discussion posts, we concluded in class that these concerns were either impossible to solve or not that big of a deal; Alive is one tool in the compiler developer’s toolbox, useful for debugging with its counterexample generation, and useful in that it provides guarantee of some correctness. Even if it’s not the magic bullet to automatically solve the problem of compiler correctness all in one go, it’s still a useful tool and has been in use in the LLVM community since its release.
+
+
+## Discussion Questions, taken from the Github conversations about the article:
+
+* Many compiler passes are not as simple as peephole optimizations, spanning basic blocks if not function calls. Alive does not support control flow analysis, but could a similar solver-based architecture be used to verify control flow modifying passes? How?
+* To what extent can we trust SMT solvers to check correctness of program transformations? It seems Alive defers the proof obligation to the SMT solver, although it is well-known that SMT solvers have bugs — there’s a [line of research](https://testsmt.github.io/) showing that popular SMT solvers have >500 soundness bugs. With this in mind, is it still ok to rely on SMT solvers? If not, how they could replace SMT solvers? They do note at the end (and parts within) that SMT solvers are expensive and some of their constraints grow uncomfortably superlinearly.
+* It's good for the developers that Alive is designed to resemble LLVM as much as possible, but does it make it hard to design or limit the expressiveness of automated verification tool like Alive in some way?
+* [Optional] What’s the best way to ensure correctness for sources with floats and aggregate types? Alive currently does not implement these. [Alive FP]
```

Comment:
What does "[Optional]" mean here? And why does this line have "[Alive FP]" at the end? (Maybe this section just needs some tweaking for this context, i.e., the externally-facing blog post instead of the internally-facing class discussion.)